File: usefull_functions.py
#
# Usefull Functions
#

def get_magnitude(From, To, Classification, Magnitudes):
    # Index distance is the distance (in integer) from (From) to (To) values inside the ORDER list
    index_distance = [Classification.index(From), Classification.index(To)]

    if index_distance[0] > index_distance[1]: # Make sure the value doesnt become something like (2, 1) instead of the usual (1, 2) since (2, 1) doesnt slice a array properly
        index_distance = [index_distance[1], index_distance[0]]

    magnitude = 1

    for value_magnitude in Magnitudes[index_distance[0]:index_distance[1]]:
        magnitude *= value_magnitude

    return magnitude

def multiply_by_magnitude(order, magnitudes, Value, From, To):
    # Check if the user is trying to convert a large unit to a small unit or vice versa
    direction = order.index(To) - order.index(From)

    if direction == 0:
        # User is trying to convert ms to ms or seconds to seconds and vice versa
        return Value # Value doesnt change since from and to are the same

    # Magnitude is the value we have to multiply/divide the value with so that we will get the desired value
    elif direction < 0:
        magnitude = get_magnitude(From, To, order, magnitudes)
        return round((Value * magnitude), 5)
    elif direction > 0:
        magnitude = get_magnitude(From, To, order, magnitudes)
        return round((Value / magnitude), 5)

def check_for_invalid_parameters(order, Value, From, To):
    if not type(Value) in [int, float]: raise ValueError('Invalid value:', Value)
    if not order.__contains__(From): raise ValueError('Invalid unit of measurement:', From)
    if not order.__contains__(To): raise ValueError('Invalid unit of measurement:', To)

File: rust/src/usefull_functions.rs
// 
// Usefull Functions
//

pub fn get_magnitude(from: &str, to: &str, classification: Vec<&str>, magnitudes: Vec<&f64>) -> f64 {
    // Index distance is the distance (in integer) from (From) to (To) values inside the ORDER list
    let index_distance = [index_of(classification.as_slice(), &from), index_of(classification.as_slice(), &to)];

    let mut magnitude: f64 = 1.0;

    for &value_magnitude in &magnitudes[(index_distance[0].unwrap() as usize)..(index_distance[1].unwrap() as usize)] {
        magnitude *= value_magnitude;
    }

    return magnitude
}


pub fn index_of<T: PartialEq>(vector: &[T], item: &T) -> Option<usize> {
    return vector.iter().position(|x| x == item)
}

pub fn multiply_by_magnitude(order: Vec<&str>, magnitudes: Vec<&f64>, value: f64, from: &str, to: &str) -> f64 {
    // Check if the user is trying to convert a large unit to a small unit or vice versa
    let startindex = index_of(&order, &from).unwrap();
    let endindex = index_of(&order, &to).unwrap();
    let direction = endindex as i32 - startindex as i32;

    if direction == 0 {
        // User is trying to convert the same unit, so value doesn't change
        return value;
    } else if direction < 0 {
        let magnitude = get_magnitude(to, from, order, magnitudes);
        return value * magnitude;
    } else {
        let magnitude = get_magnitude(from, to, order, magnitudes);
        return value / magnitude;
    }
}

pub fn check_for_invalid_parameters(order: &[&str], from: &str, to: &str) {
    if !order.contains(&from) { panic!("Invalid unit of measurement: {}", from); }
    if !order.contains(&to) { panic!("Invalid unit of measurement: {}", to); }
}

